cirq/circuits/circuit_test.py

```diff
@@ -4432,3 +4432,166 @@ def _circuit_diagram_info_(self, args):
 
     # Escaping Special Characters in Qubit names.
     assert '|c&gt;' in circuit._repr_html_()
+
+
+def test_tetris_concat():
+    a, b = cirq.LineQubit.range(2)
+    empty = cirq.Circuit()
+
+    assert cirq.Circuit.tetris_concat(empty, empty) == empty
+    assert cirq.Circuit.tetris_concat() == empty
+    assert empty.tetris_concat(empty) == empty
+    assert empty.tetris_concat(empty, empty) == empty
+
+    ha = cirq.Circuit(cirq.H(a))
+    hb = cirq.Circuit(cirq.H(b))
+    assert ha.tetris_concat(hb) == ha.zip(hb)
+
+    assert ha.tetris_concat(empty) == ha
+    assert empty.tetris_concat(ha) == ha
+
+    hac = cirq.Circuit(cirq.H(a), cirq.CNOT(a, b))
+    assert hac.tetris_concat(hb) == hac + hb
+    assert hb.tetris_concat(hac) == hb.zip(hac)
+
+    zig = cirq.Circuit(cirq.H(a), cirq.CNOT(a, b), cirq.H(b))
+    assert zig.tetris_concat(zig) == cirq.Circuit(
+        cirq.H(a), cirq.CNOT(a, b), cirq.Moment(cirq.H(a), cirq.H(b)), cirq.CNOT(a, b), cirq.H(b)
+    )
+
+    zag = cirq.Circuit(cirq.H(a), cirq.H(a), cirq.CNOT(a, b), cirq.H(b), cirq.H(b))
+    assert zag.tetris_concat(zag) == cirq.Circuit(
+        cirq.H(a),
+        cirq.H(a),
+        cirq.CNOT(a, b),
+        cirq.Moment(cirq.H(a), cirq.H(b)),
+        cirq.Moment(cirq.H(a), cirq.H(b)),
+        cirq.CNOT(a, b),
+        cirq.H(b),
+        cirq.H(b),
+    )
+
+    space = cirq.Circuit(cirq.Moment()) * 10
+    f = cirq.Circuit.tetris_concat
+    assert len(f(space, ha)) == 10
+    assert len(f(space, ha, ha, ha)) == 10
+    assert len(f(space, f(ha, ha, ha))) == 10
+    assert len(f(space, ha, stop_at_first_alignment=True)) == 10
+    assert len(f(space, ha, ha, ha, stop_at_first_alignment=True)) == 12
+    assert len(f(space, f(ha, ha, ha, stop_at_first_alignment=True))) == 10
+
+    # L shape overlap (vary c1).
+    assert 7 == len(
```

Comment:
These almost look like parametrized tests...which is a better pattern than multiple assertions, if there are failures, it is easier to understand what broke compared to whack-a-mole with the multiple assertions. Would you mind rewriting it to parameterized tests?

Reply:
My rule of thumb writing unit tests is that they should be very literal; minimize abstractions. Abstractions are chances to make mistakes and to obscure what is being tested. This does tend to give my tests a repetitive nature, and of course I break the rule all the time. This case is particularly relevant, because the reason I am being so explicit is because I am checking for off by one errors in min/max logic. If I write the test in a parameterized way, I will have to do the assertion using min/max logic prone to off-by-one errors. This correlates the test being wrong with the code being wrong, which is very bad.

Generally I parameterize when I have a large number of cases, or a variable number of cases (e.g. "all gates"), or the property being tested is decoupled from the values (e.g. "addition commutes"). Here the number of cases is small and fixed (5), and the cases are very tightly coupled with what is being tested (does the pole hit the backstop or not).